lib/sax.js

```diff
@@ -554,6 +554,33 @@ function parseDCC(source,start,domBuilder,errorHandler){//sure start with '<!'
 			domBuilder.endCDATA() 
 			return end+3;
 		}
+		if(source.substr(start+2,6) == 'ENTITY'){
+			var end = source.indexOf('>', start+8);
+			var chunk = source.substring(start+8, end);
+			var match = chunk.match(/\s+(\w+)\s+(["'])/);
+			if (!match) {
+				errorHandler.error("Malformed internal entity declaration");
+				return -1;
+			}
+			var declStart = start+8+match[0].length;
+			var name = match[1];
+			var delim = match[2];
+			var delimEnd = source.indexOf(delim, declStart);
+			if (delimEnd > end) {
+				end = source.indexOf('>', delimEnd);
+			}
+			var value = source.substring(declStart, delimEnd);
+			// NOTE: This value is not further processed, but treated as PCDATA.
+			// If recursive processing is implemented, ENSURE to avoid an XML Bomb
+			// attack vector!
+			domBuilder.internalEntityDecl(name, value);
+			var nextTagStart = source.indexOf('<', end);
+			var dtdEnd = source.indexOf(']>', end);
+			if (dtdEnd < nextTagStart) {
+				domBuilder.endDTD();
+			}
```

Comment:
I just found the following section of the spec:
https://www.w3.org/TR/2006/REC-xml11-20060816/#sec-predefined-ent

> All XML processors must recognize these entities whether they are declared or not. For interoperability, valid XML documents should declare these entities, like any others, before using them. If the entities lt or amp are declared, they must be declared as internal entities whose replacement text is a character reference to the respective character (less-than sign or ampersand) being escaped; the double escaping is required for these entities so that references to them produce a well-formed result. If the entities gt, apos, or quot are declared, they must be declared as internal entities whose replacement text is the single character being escaped (or a character reference to that character; the double escaping here is optional but harmless). For example:

```
<!ENTITY lt     "&#38;#60;">
<!ENTITY gt     "&#62;">
<!ENTITY amp    "&#38;#38;">
<!ENTITY apos   "&#39;">
<!ENTITY quot   "&#34;">
```

Does the current implementation support this? (We should add a test if it's not already there.)
I didn't look into the tests yet, but I wonder if it could make sense to at least do a single replacement run as part of `domBuilder.internalEntityDecl` since it has access to the existing `entityMap`. Or would that already open up the "can of worms"?

Reply:
Ah, no, this specific recommended re-declaration isn't technically supported (and there is apparently no test for it). Of course these are predefined so they are already available (and I cannot recall seeing this re-declaration in practise, at least not internally in documents).

As I read this, it seems that while re-declarations are allowed to issue warnings, these specific characters (being the `XML_ENTITIES`) are allowed to be re-declared by these exact values. That could be hard-coded, I guess, now or later on, to avoid the then improper warning if this is done.

And yes, I believe doing a replacement run on the entity values could open up that can. The [Billion laughs attack](https://en.wikipedia.org/wiki/Billion_laughs_attack) doesn't rely upon recursion, but upon exponential growth, which this would enable. It should then be combined with e.g. monitoring of value growth, such as by carefully counting expansions, as libxml2 does with its entity reference loop detection (and corresponding `--huge` option to `xmllint`).

Reply:
Thank you for elaborating on this topic, very much appreciated.
After reading the Wikipedia article it became obvious.

I fully agree that there is no need for implementing this "exception" of the "first definition is binding" rule for XML entities, since the only allowed option is to redefine them to the same value, so in the worst case there is a warning about this not being supported, but the values still being correct.

I just started to think about (X)HTML and entities.
- XHTML both get's the html entities and is otherwise working like XML, so that shouldn't be a problem.
- Since the living HTML spec clearly defines what the doctype can contain:
https://html.spec.whatwg.org/multipage/syntax.html#the-doctype
we should check what the browser does if it is provided a doctype with entities and the html mime type, and act accordingly.

Reply:
Sorry, my confusion, currently xmldom basically only knows about XHTML and treats HTML in the "same way".
So this will be part of #338